tokio-postgres/src/copy_both.rs

```diff
@@ -0,0 +1,280 @@
+///       v v
+///   ReadingCopy
+///        |
+///        v
+///     Reading
+/// ```
+#[derive(Clone, Copy)]
+enum DuplexState {
+    /// Initial state where CopyData messages can go in both directions
+    CopyBoth,
+    /// The server->client stream is closed and we're in CopyIn mode
+    CopyIn,
+    /// The client->server stream is closed and we're in CopyOut mode
+    CopyOut,
+    /// Both directions are closed and we're waiting for the first CommandComplete message
+    ReadingCopy,
+    /// Both directions are closed and we're waiting for the final CommandComplete message
+    Reading,
+}
+
+pin_project! {
+    /// A duplex stream for consuming streaming replication data.
+    ///
+    /// The stream side *must* be consumed even if not required in order to process the messages
+    /// coming from the server. If it is not, `Sink::close` may hang forever waiting for the stream
+    /// messages to be consumed.
+    ///
+    /// The copy should be explicitly completed via the `Sink::close` method to ensure all data is
+    /// flushed and errors are received.
+    pub struct CopyBothDuplex<T> {
+        #[pin]
+        message_sender: mpsc::Sender<FrontendMessage>,
+        #[pin]
+        error_sender: Option<oneshot::Sender<()>>,
+        responses: Responses,
+        buf: BytesMut,
+        state: DuplexState,
+        #[pin]
+        _p: PhantomPinned,
+        _p2: PhantomData<T>,
+    }
+}
+
+impl<T> Stream for CopyBothDuplex<T> {
+    type Item = Result<Bytes, Error>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        use DuplexState::*;
+
+        let mut this = self.project();
+        let unexpected = Poll::Ready(Some(Err(Error::unexpected_message())));
+
+        loop {
+            match ready!(this.responses.poll_next(cx)?) {
+                Message::CopyData(body) => {
+                    return match this.state {
+                        CopyBoth | CopyOut => Poll::Ready(Some(Ok(body.into_bytes()))),
+                        _ => unexpected,
+                    };
+                }
+                // The server->client stream is done
+                Message::CopyDone => {
+                    *this.state = match this.state {
+                        CopyBoth => CopyIn,
+                        CopyOut => ReadingCopy,
+                        _ => return unexpected,
+                    };
+                }
+                // The server indicated an error, terminate both sides
+                Message::ErrorResponse(error) => {
+                    *this.state = match this.state {
+                        ReadingCopy | Reading => return unexpected,
+                        _ => ReadingCopy,
+                    };
+                    // Indicate to CopyBothReceiver to produce a Sync message instead of CopyDone
+                    let _ = this.error_sender.take().unwrap().send(());
+                    return Poll::Ready(Some(Err(Error::db(error))));
```

Comment:
Does this assume that the error happened during `START_REPLICATION`? What about an error that happened during the stream; shouldn't we still send a `CopyDone`?